Apps/FlightSoftware/Watchdog2/src/uart.c

```diff
@@ -0,0 +1,393 @@
+
+        /* clear UART_A0 receive flag */
+        /* we are done here */
+        UCA0IFG &= ~UCRXIFG;
+        break;
+    default: /* some other possibilities */
+        break;
+    }
+}
+
+__volatile int is_escaped = 0;
+__volatile int has_started = 0;
+
+/* UCA1 interrupt handler (watchdog SLIP) */
+#if defined(__TI_COMPILER_VERSION__) || defined(__IAR_SYSTEMS_ICC__)
+#pragma vector=EUSCI_A1_VECTOR
+__interrupt void USCI_A1_ISR(void) {
+#elif defined(__GNUC__)
+void __attribute__ ((interrupt(EUSCI_A1_VECTOR))) USCI_A1_ISR (void) {
+#else
+#error Compiler not supported!
+#endif
+
+    unsigned char rcv;
+
+    // TODO: buffer wraparound seems weird right now.
+
+    /* two possibilities; rx or tx */
+    switch(__even_in_range(UCA1IV, USCI_UART_UCTXCPTIFG)) {
+    case USCI_UART_UCTXIFG: /* transmitted byte successfully */
+        /* decrement the number of bytes used */
+        uart1tx.used--;
+
+        if (uart1tx.used == 0) {
+            /* done sending after this byte; clear IFG */
+            UCA1IE &= ~UCTXIE;
+            /* unlock the buffer */
+            uart1tx.locked = 0;
+        }
+
+        /* send the next byte */
+        UCA1TXBUF = uart1tx.buf[uart1tx.idx++];
+        if (uart1tx.idx >= BUFFER_SIZE) uart1tx.idx = 0;
+        break;
+    case USCI_UART_UCRXIFG: /* received new byte */
+        /* get the received character */
+        rcv = UCA1RXBUF;
+
+        if (!has_started) {
+            if (rcv == SLIP_END) {
+                /* start now */
+                has_started = 1;
+                goto end_rx;
+            } else {
+                goto end_rx;
+            }
+        }
+
+        /* deal with escaped characters */
+        if (is_escaped && rcv == SLIP_ESC_END) {
+            /* push the escaped character */
+            uart1rx.buf[uart1rx.idx++] = SLIP_END;
+            is_escaped = 0;
+            goto end_rx;
+        } else if (is_escaped && rcv == SLIP_ESC_ESC) {
+            /* push the escaped character */
+            uart1rx.buf[uart1rx.idx++] = SLIP_ESC;
+            is_escaped = 0;
+            goto end_rx;
+        }
+
+        /* regular case (not escaped character) */
+        switch (rcv) {
+        case SLIP_END:
+            /* done reading; skip storing the end byte, and signal to the main loop that we are done */
+            loop_flags |= FLAG_UART1_RX_PACKET;
+            has_started = 0;
+            // exit LPM
+            __bic_SR_register(DEFAULT_LPM);
+            break;
+        case SLIP_ESC:
+            /* about to start escape sequence; skip storing this byte */
+            is_escaped = 1;
+            break;
+        default:
+            /* regular byte */
+            uart1rx.buf[uart1rx.idx++] = rcv;
+            break;
+        }
+
+        /* clear UART_A1 receive flag */
+        /* we are done here */
+    end_rx:
+        UCA1IFG &= ~UCRXIFG;
+        if (uart1rx.idx >= BUFFER_SIZE) {
+            uart1rx.idx -= BUFFER_SIZE;
+        }
+        break;
```

Comment:
Gross, goto. This could be rewritten as follows to eliminate it:
```
        /* get the received character */
        rcv = UCA1RXBUF;

        if (!has_started) {
            if (rcv == SLIP_END) {
                /* start now */
                has_started = 1;
            }
        } else if (is_escaped) { /* deal with escaped characters */
            /* push the escaped character */
            if (rcv == SLIP_ESC_END) { 
                uart1rx.buf[uart1rx.idx++] = SLIP_END;
            } else if (rcv == SLIP_ESC_ESC) {
                uart1rx.buf[uart1rx.idx++] = SLIP_ESC;
            }
       
            is_escaped = 0;
        } else { /* regular case (not escaped character) */
            switch (rcv) {
            case SLIP_END:
                /* done reading; skip storing the end byte, and signal to the main loop that we are done */
                loop_flags |= FLAG_UART1_RX_PACKET;
                has_started = 0;
                // exit LPM
                __bic_SR_register(DEFAULT_LPM);
                break;
            case SLIP_ESC:
                /* about to start escape sequence; skip storing this byte */
                is_escaped = 1;
                break;
            default:
                /* regular byte */
                uart1rx.buf[uart1rx.idx++] = rcv;
                break;
            }
        }

        /* clear UART_A1 receive flag */
        /* we are done here */
        UCA1IFG &= ~UCRXIFG;
        if (uart1rx.idx >= BUFFER_SIZE) {
            uart1rx.idx -= BUFFER_SIZE;
        }
        break;
```